### PR TITLE
VSCode Plugin initial implementation

### PR DESCRIPTION
**Pull request by hsubox76**
This adds the code for the VS Code extension (or plugin) as a subdirectory in the firebase-tools repo. This subdirectory will be ignored by firebase-tools top level `build` and `test` scripts and has its own build scripts in its own `package.json`.

Please ignore files outside of `firebase-vscode/` as these are in a pending PR soon to be merged: https://github.com/firebase/firebase-tools/pull/5667

For an **overview** of the design decisions see the TDD (WIP): https://docs.google.com/document/d/1wwFywuFt9JL07kKqM4PAp077OkWGH0wLb5Hevw6nO04/edit?resourcekey=0-neCwj0srvoswM8mBdomCsQ# (internal)

For a **detailed description** of the project structure, files, configs, etc see: https://docs.google.com/document/d/1xngcpsJaPyTUgnYHP5J1reu5zzNyAJIQyDOaVrXwCbg/edit# (internal)

Would welcome any checks or stress tests to ensure it won't affect CLI builds or CI.

### CODE REVIEW

File: firebase-vscode/webviews/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    "moduleResolution": "Node",
+    "target": "ES2020",
+    "lib": ["ES2020", "DOM"],
+    "esModuleInterop": true,
+    "jsx": "react",
+    "sourceMap": true,
+		"rootDirs": ["./", "../../src", "../common"],
+    "strict": false /* enable all strict type-checking options */
+    /* Additional Checks */
+    // "noImplicitReturns": true, /* Report error when not all code paths in function return a value. */
```
Comment (christhompsongoogle) [inline firebase-vscode/webviews/tsconfig.json:13]: nit: please uncomment or reformat

File: src/monospace/index.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    // wait and call back
+    await new Promise((res) => setTimeout(res, POLL_USER_RESPONSE_MILLIS));
+
+    // TODO: decide how long to ultimately wait before declaring
+    // that the user is never going to respond.
+
+    return pollAuthorizedProject(rid);
+  }
+
+  // FIXME: This is not being reached as the process exits before a new call is made
```
Comment (christhompsongoogle) [inline src/monospace/index.ts:75]: nit replace FIXME

File: firebase-vscode/src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import * as fs from "fs";
+import { FirebaseRC } from "../../src/firebaserc";
+
+export async function writeFirebaseRCFile(
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/utils.ts:4]: It is used in workflow.ts but can look into if the CLI one will work.

File: firebase-vscode/package.json [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+          "type": "string",
+          "default": "",
+          "description": "Firebase RC folder"
+        }
+      }
+    },
+    "viewsContainers": {
+      "activitybar": [
+        {
+          "id": "firebase2",
```
Comment (hsubox76) [inline firebase-vscode/package.json:33]: All these "firebase2"s were supposed to be temporary - it was conflicting with the hackweek project embedded into Monospace which used the same ID. Need to revert these back to `firebase`.

**Review by joehan — COMMENTED** [review]
Got started on this & made it to cli.ts, will come back to this again

File: firebase-vscode/common/messaging/protocol.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  /** Runs `firebase deploy` for hosting. */
+  hostingDeploy(target: string): void;
+
+  /** fetches a list of folders in the user's workspace. */
+  getWorkspaceFolders(): void;
+
+  /** get selected project either from firebaserc or last cached value (or workspace file) */
+  getSelectedProject(): void;
+
+  /** Fetches the entire firebase rc config file. If the file doesn't exist, then it will return a default value. */
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/common/messaging/protocol.ts]: Clarified.

File: firebase-vscode/common/messaging/protocol.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+  /** Called when a new project is selected */
+  notifyProjectChanged(projectId: string): void;
+
+  /**
+   * This can potentially call multiple webviews to notify of user selection.
+   */
+  notifyUserChanged(email: string): void;
+
+  notifyHostingFolderReady(projectId: string, folderPath: string): void;
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/common/messaging/protocol.ts:82]: Added.

File: firebase-vscode/package.json [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      "firebase2": [
+        {
+          "type": "webview",
+          "id": "firebase2.sidebarView",
+          "name": "Firebase"
+        }
+      ]
+    },
+    "commands": [
+      {
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/package.json]: Oops. I guess the entire commands section is unused for now.

File: firebase-vscode/package.json [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    "dev:sidebar": "npm run copyfiles && webpack --config webpack.dev.js --config-name sidebar",
+    "watch": "node scripts/swap-pkg.js vsce && npm run copyfiles && webpack --config webpack.dev.js --watch",
+    "build": "npm run copyfiles && webpack --config webpack.prod.js --devtool hidden-source-map",
+    "build:extension": "webpack --config webpack.prod.js --config-name extension",
+    "build:sidebar": "npm run copyfiles && webpack --config webpack.prod.js --config-name sidebar",
+    "pretest": "npm run build && npm run lint",
+    "lint": "eslint src --ext ts",
+    "test": "node ./dist/test/runTest.js"
+  },
+  "devDependencies": {
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/package.json:81]: I moved the ones I think are appropriate, let me know.

File: firebase-vscode/scripts/swap-pkg.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+process.argv.forEach(arg => {
+  if (arg === 'vsce' || arg === 'monospace') {
+    target = arg;
+  }
+});
+
+if (target === 'vsce') {
+  delete pkg.extensionDependencies;
+  console.log(JSON.stringify(pkg, null, 2));
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/scripts/swap-pkg.js]: Oops, these were debugging logs accidentally left in. It would be good to put a useful message though so added.

File: firebase-vscode/src/cli.ts [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+  const options = { ...currentOptions };
+  if (!options.project) {
+    return [];
+  }
+  // TODO: handle multiple hosting configs
+  if (!(firebaseJSON.hosting as HostingSingle).site) {
+    (firebaseJSON.hosting as HostingSingle).site =
+      await getDefaultHostingSite(options);
+  }
+  pluginLogger.debug('Calling listChannels with params', options.project, (firebaseJSON.hosting as HostingSingle).site);
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/cli.ts]: Hm I don't understand why prettier won't automatically handle this.

File: firebase-vscode/src/logger-wrapper.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
+
+import { logger as cliLogger } from "../../src/logger";
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/logger-wrapper.ts:2]: No, let me know if you see any console.logs, other than the one in that build util script.

File: firebase-vscode/src/sidebar.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  context.subscriptions.push(
+    window.registerWebviewViewProvider(
+      MonospaceSidebarViewProvider.viewType,
+      provider
+    )
+  );
+  return provider;
+}
+
+class MonospaceSidebarViewProvider implements WebviewViewProvider {
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/sidebar.ts]: Nope, this was overlooked, fixed.

File: firebase-vscode/src/stubs/empty-class.js [hunk 13]
```diff
@@ -0,0 +1,3 @@
+class Noop {}
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/stubs/empty-class.js:1]: See webpack.common.js and the second doc in the description - need to no-op some CLI dependencies that break the extension/webpack, without touching the CLI code.

File: firebase-vscode/src/stubs/marked.js [hunk 14]
```diff
@@ -0,0 +1,5 @@
+function marked() {}
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/stubs/marked.js:1]: Need this stub to prevent errors, needs to match the signature as called by the CLI (it is called as a function).

File: firebase-vscode/src/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/utils.ts:6]: Will also have to use `loadRC` and the `RC` class - probably the best way - added a todo since it will be somewhat extensive.

File: firebase-vscode/tsconfig.json [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+		"outDir": "dist",
+		"lib": [
+			"ES2020"
+		],
+		"jsx": "react",
+		"sourceMap": true,
+		"rootDirs": ["src", "../src", "common"],
+		"strict": false   /* enable all strict type-checking options */
+		/* Additional Checks */
+		// "noImplicitReturns": true, /* Report error when not all code paths in function return a value. */
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/tsconfig.json]: These are from the vscode template, no need to keep them.

File: src/monospace/index.ts (hunk 2 above)
Comment (hsubox76) [reply src/monospace/index.ts:75]: Needs to be changed in https://github.com/firebase/firebase-tools/pull/5667

**Review by joehan — COMMENTED** [review]
Saw you addressed the first round, sending the comments i had queued up & taking a look at the changes

**Review by joehan — COMMENTED** [review]
Saw you addressed the first round, sending the comments i had queued up & taking a look at the changes

File: firebase-vscode/common/messaging/types.d.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
+import { Channel } from "../hosting/api";
+
+export interface Message {
+  message: string;
+  data: any[];
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/common/messaging/types.d.ts:5]: Do you know of a way to do this? I guess the type for data would have to be an array where the member type is all the params in the functions in WebviewToExtension and ExtensionToWebview interfaces in common/messaging/protocol.ts, is there any way to access those with TypeScript? I see we can get the keys, and we can get the functions as the values, but is there a way to get the params of those functions?

File: firebase-vscode/src/cli.ts [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+import { Options } from "../../src/options";
+import { currentOptions, getCommandOptions } from "./options";
+import { setInquirerOptions } from "./stubs/inquirer-stub";
+import * as vscode from "vscode";
+import { ServiceAccount } from "../common/types";
+import { listChannels } from "../../src/hosting/api";
+import { ChannelWithId } from "../common/messaging/types";
+import { setEnabled } from "../../src/experiments";
+import { pluginLogger } from "./logger-wrapper";
+import { inspect } from "util";
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/cli.ts:26]: I can do this manually here, there is an eslint plugin that could do it automatically if we want to add it globally to firebase-tools (probably not in this PR): https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md

File: firebase-vscode/src/extension-broker.ts [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+  sendMessage(message: string, data: any[]): void {
+    this.webviews.forEach((webview) => {
+      webview.postMessage({ message, data });
+    });
+  }
+
+  registerReceiver(receiver: any) {
+    const webview = receiver as Webview;
+    this.webviews.push(webview);
+    webview.onDidReceiveMessage((data: any) => {
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/extension-broker.ts]: See other comment, I think it would have to be a clever generic? I can't think of how to do it off the bat.

File: firebase-vscode/src/cli.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (joehan) [reply firebase-vscode/src/cli.ts]: Good find, I'll give that a try

File: firebase-vscode/common/messaging/broker.ts [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+  registerReceiver(receiver: any): void { }
+
+  addListener(message: string, cb: Listener): void {
+    if (!this.listeners[message]) {
+      this.listeners[message] = { listeners: [] };
+    }
+    this.listeners[message].listeners.push(cb);
+  }
+
+  executeListeners(data: Message) {
```
Comment (joehan) [inline firebase-vscode/common/messaging/broker.ts:21]: Naming is a bit hard to follow here, since `data` is a message and message has a field `data`

File: firebase-vscode/common/messaging/broker.ts [hunk 20]
```diff
@@ -0,0 +1,87 @@
+import { MessageParamsMap } from "./protocol";
+import { Listener, Message, MessageListeners } from "./types";
+import { Webview } from "vscode";
+
+const isObject = (val: any): boolean => typeof val === "object" && val !== null;
+
+type Receiver = {} | Webview;
+
+export abstract class Broker<
```
Comment (hsubox76) [inline firebase-vscode/common/messaging/broker.ts:9]: I messed around trying to tighten up the types, let me know if this is a good idea or just a big mess.

Comment (christhompsongoogle) [reply firebase-vscode/common/messaging/broker.ts:9]: We can always loosen them later IMO - best to start strict than the other way around

File: firebase-vscode/common/messaging/protocol.ts [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+
+  /**
+   * Notify webview of initial discovery or change in firebase.json or
+   * .firebaserc
+   */
+  notifyFirebaseConfig: { firebaseJson: FirebaseConfig, firebaseRC: FirebaseRC };
+
+}
+
+export type MessageParamsMap = WebviewToExtensionParamsMap | ExtensionToWebviewParamsMap;
```
Comment (christhompsongoogle) [inline firebase-vscode/common/messaging/protocol.ts:120]: Nit: NL

File: firebase-vscode/src/html-scaffold.ts [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+    Uri.joinPath(extensionUri, "resources/Monicons.woff")
+  );
+  const codiconsUri = webview.asWebviewUri(
+    Uri.joinPath(extensionUri, "resources/dist/codicon.css")
+  );
+  // Use a nonce to only allow a specific script to be run.
+  const nonce = getNonce();
+
+  return `<!DOCTYPE html>
+			<html lang="en">
```
Comment (christhompsongoogle) [inline firebase-vscode/src/html-scaffold.ts:24]: Nit: tabs -> spaces and formatting to fit. If we start on the next line we can get back ~10 characters or so

File: firebase-vscode/webviews/SidebarApp.tsx [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+   * empty array - finished checking, no users logged in
+   * non-empty array - contains logged in users
+   */
+  const [allUsers, setAllUsers] = useState<Array<
+    ServiceAccountUser | User
+  > | null>(null);
+  const [isHostingOnboarded, setHostingOnboarded] = useState<boolean>(false);
+
+  useEffect(() => {
+    console.log("loading SidebarApp component");
```
Comment (christhompsongoogle) [inline firebase-vscode/webviews/SidebarApp.tsx:34]: A handful of console.log in this file

File: firebase-vscode/webviews/components/AccountSection.scss [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+
+  &-icon {
+    margin-right: 8px;
+  }
+
+  &-project {
+    display: flex;
+    flex-direction: column;
+  }
+}
```
Comment (christhompsongoogle) [inline firebase-vscode/webviews/components/AccountSection.scss:20]: NL

File: firebase-vscode/webviews/components/ui/PanelSection.scss [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+  color: var(--vscode-descriptionForeground);
+
+  .panelExpandoIcon {
+    transition: transform 0.1s ease;
+  }
+
+  &:not(.isExpanded) .panelExpandoIcon {
+    transform: rotate(-90deg);
+  }
+}
```
Comment (christhompsongoogle) [inline firebase-vscode/webviews/components/ui/PanelSection.scss:20]: NL

File: firebase-vscode/webviews/components/ui/PanelSection.tsx [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+        <>
+          {title && <Spacer size="medium" />}
+          {children}
+          <Spacer size="xlarge" />
+          {!isLast && <VSCodeDivider style={{ width: "100vw" }} />}
+        </>
+      )}
+    </>
+  );
+}
```
Comment (christhompsongoogle) [inline firebase-vscode/webviews/components/ui/PanelSection.tsx:45]: NL

File: firebase-vscode/webviews/globals/index.scss [hunk 27]
```diff
@@ -0,0 +1,12 @@
...
+
+body {
+  background-color: transparent;
+  cursor: default;
+}
+
+:global #root {
+  display: flex;
+  flex-direction: column;
+}
```
Comment (christhompsongoogle) [inline firebase-vscode/webviews/globals/index.scss:12]: NL

File: firebase-vscode/webviews/sidebar.entry.scss [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+
+  &-project {
+    display: flex;
+    flex-direction: column;
+  }
+  
+  &-test {
+    display: flex;
+  }
+}
```
Comment (christhompsongoogle) [inline firebase-vscode/webviews/sidebar.entry.scss:54]: NL

File: firebase-vscode/webpack.common.js [hunk 29]
```diff
@@ -0,0 +1,199 @@
...
+        }
+      ],
+    })
+  ],
+  infrastructureLogging: {
+    level: "log", // enables logging required for problem matchers
+  },
+};
+
+const makeWebConfig = (entryName) => {
```
Comment (christhompsongoogle) [inline firebase-vscode/webpack.common.js:109]:
```suggestion
function makeWebConfig(entryName: string): SomeConfig {
```

File: firebase-vscode/common/messaging/broker.ts (hunk 19 above)
Comment (hsubox76) [reply firebase-vscode/common/messaging/broker.ts]: `Message` also has a field `message`, which doesn't help. I renamed Message.message to Message.command and this to `message: Message`.

File: firebase-vscode/src/html-scaffold.ts (hunk 22 above)
Comment (hsubox76) [reply firebase-vscode/src/html-scaffold.ts:24]: Done. I think.

File: firebase-vscode/webviews/SidebarApp.tsx (hunk 23 above)
Comment (hsubox76) [reply firebase-vscode/webviews/SidebarApp.tsx:34]: I made a logger util for webviews as well that passes the messages to the extension process for debug logging.

**Review by christhompsongoogle — APPROVED** [review]

File: firebase-vscode/.eslintrc.json [hunk 30]
```diff
@@ -0,0 +1,24 @@
+{
+    "root": true,
+    "parser": "@typescript-eslint/parser",
+    "parserOptions": {
+        "ecmaVersion": 6,
+        "sourceType": "module"
+    },
+    "plugins": [
+        "@typescript-eslint"
+        // "react"
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase-vscode/.eslintrc.json:10]: I'd say go for it. It's the least bad option

File: firebase-vscode/LICENSE
Comment (yuchenshi) [inline firebase-vscode/LICENSE:1]: BLOCKING: This file cannot be empty. Please either copy & paste the top-level LICENSE or somehow package and release that file as part of the releasing process if you want to avoid duplication.

File: firebase-vscode/webviews/components/ui/PanelSection.tsx [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+  children,
+  isLast,
+}: {
+  title?: ReactNode;
+  children: ReactNode;
+  isLast?: boolean;
+}) {
+  let [isExpanded, setExpanded] = useState(true);
+  if (!title) {
+    isExpanded = true;
```
Comment (yuchenshi) [inline firebase-vscode/webviews/components/ui/PanelSection.tsx:20]: This is confusing -- consider just changing the condition on Line 35 instead.

File: firebase-vscode/package.json [hunk 32]
```diff
@@ -0,0 +1,110 @@
+{
+  "name": "firebase-vscode",
+  "displayName": "firebase-vscode",
+  "publisher": "firebase",
+  "description": "",
```
Comment (yuchenshi) [inline firebase-vscode/package.json:5]: nit: Anything good we can put in here?

**Review by yuchenshi — APPROVED** [review]

File: firebase-vscode/src/extension.ts [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+  WebviewToExtensionParamsMap,
+  vscode.Webview
+>(new ExtensionBroker());
+
+// This method is called when your extension is activated
+// Your extension is activated the very first time the command is executed
+export function activate(context: vscode.ExtensionContext) {
+  // Use the console to output diagnostic information (console.log) and errors (console.error)
+  // This line of code will only be executed once when your extension is activated
+  console.log(
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/extension.ts:25]: Removed the extra comments, and replaced the console.log with a pluginLogger.debug to log that the extension is being activated.

File: firebase-vscode/webpack.common.js [hunk 34]
```diff
@@ -0,0 +1,199 @@
...
+            },
+            {
+              search: /(\.|\.\.)[\.\/]+schema/g,
+              replace: "./schema",
+            },
+            {
+              search: /Configstore\(pkg\.name\)/g,
+              replace: "Configstore('firebase-tools')",
+            },
+            // TODO(hsubox76): replace with something more robust
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/webpack.common.js:80]: I'm trying to avoid changing any CLI code at the moment, this is a stopgap for now to prevent complete breakage. I'll eventually have to go into the CLI code (src/deploy/lifecycleHooks.ts) and figure out a way to bundle `cross-env-shell.js` and run it so that the predeploy can work in windows, and then somehow call the bundled version from the extension. In this initial PR I'm trying to avoid making any changes that could potentially affect the functionality of the CLI.

File: firebase-vscode/webviews/components/ui/PanelSection.tsx (hunk 31 above)
Comment (hsubox76) [reply firebase-vscode/webviews/components/ui/PanelSection.tsx:20]: Yeah I don't actually understand the purpose of the conditional because line 18 defaults it to true anyway. Removed the conditional. We do need the state because it's toggleable by the user.

File: firebase-vscode/package.json (hunk 32 above)
Comment (hsubox76) [reply firebase-vscode/package.json:5]: This reminds me, I just put the name and display name as a placeholder and probably need to talk to the wider team and PM about that.

File: firebase-vscode/src/stubs/inquirer-stub.js [hunk 35]
```diff
@@ -0,0 +1,23 @@
+const inquirer = module.exports;
+
+let options = {};
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/src/stubs/inquirer-stub.js:3]: Done, let me know if that's the right way to use a Symbol.

File: firebase-vscode/src/logger-wrapper.ts [hunk 36]
```diff
@@ -0,0 +1,13 @@
...
+
+export const pluginLogger: Record<string, any> = {};
```
Comment (joehan) [inline firebase-vscode/src/logger-wrapper.ts:4]: I think this any can be LogMethod instead (exported from Logger), or at least something like  `(...args) => void`

File: firebase-vscode/src/stubs/inquirer-stub.js [hunk 37]
```diff
@@ -0,0 +1,24 @@
...
+  inquirer[optionsKey] = inquirerOptions;
+};
+
+inquirer.prompt = async (prompts) => {
+  const answers = {};
+  for (const prompt of prompts) {
+    if (inquirer[optionsKey].hasOwnProperty(prompt.name)) {
+      answers[prompt.name] = inquirer[optionsKey][prompt.name];
+    } else {
+      console.log(
```
Comment (joehan) [inline firebase-vscode/src/stubs/inquirer-stub.js:16]: Should this be switched to use the logger you implemented?

File: firebase-vscode/webviews/components/ui/PanelSection.tsx (hunk 31 above)
Comment (yuchenshi) [reply firebase-vscode/webviews/components/ui/PanelSection.tsx]: Without the extra logic, say if a user closes a panel but later the panel title is changed to empty, they will be left in a state where they can no longer see anything at all (since there's no title and no expand button)

File: firebase-vscode/src/workflow.ts [hunk 38]
```diff
@@ -0,0 +1,429 @@
...
+import { User } from "../../src/types/auth";
+import { FirebaseRC } from "../../src/firebaserc";
+import { FirebaseConfig } from "../../src/firebaseConfig";
+import { currentOptions, updateOptions } from "./options";
+import { ServiceAccountUser } from "./types";
+import { selectProjectInMonospace } from "../../src/monospace";
+import { setupLoggers, tryStringify } from "../../src/utils";
+import { pluginLogger } from "./logger-wrapper";
+import { logger } from '../../src/logger';
+import { transports, format } from "winston";
```
Comment (joehan) [inline firebase-vscode/src/workflow.ts:26]: Same suggestion as before, convention is to split external and internal imports. Totally ok if you want to wait for me to add the linter rule

File: firebase-vscode/src/workflow.ts [hunk 39]
```diff
@@ -0,0 +1,429 @@
...
+      { user: User | ServiceAccountUser }
+  ) => {
+    if (users.some((user) => user.email === requestedUser.email)) {
+      currentUserEmail = requestedUser.email;
+      broker.send("notifyUserChanged", { email: currentUserEmail });
+    }
+  });
+
+  broker.on("selectProject", async ({ email }) => {
+    let projectId;
```
Comment (joehan) [inline firebase-vscode/src/workflow.ts:251]: Optional: Some of these broker.on calls have a lot of logic and functionality tucked away in anonymous functions. It may be worth separating them out into named functions (or function generators) for cleaner organization and easier unit testing.

File: firebase-vscode/webviews/globals/web-logger.ts [hunk 40]
```diff
@@ -0,0 +1,18 @@
...
+
+for (const level of levels) {
+  tempObject[level] = (...args: string[]) =>
+    broker.send('writeLog', { level, args });
+}
+
+// Recast it now that it's populated.
+const webLogger = tempObject as WebLogger;
+
+export { webLogger };
```
Comment (joehan) [inline firebase-vscode/webviews/globals/web-logger.ts:18]: nit: newline?

**Review by joehan — APPROVED** [review]
Another round of suggestions, but nothing blocking

File: firebase-vscode/webviews/components/ui/PanelSection.tsx [hunk 41]
```diff
@@ -0,0 +1,42 @@
...
+        <button
+          aria-label={(isExpanded ? "Hide" : "Toggle") + " " + title}
+          className={cn(styles.panelExpando, isExpanded && styles.isExpanded)}
+          onClick={() => setExpanded(!isExpanded)}
+        >
+          <Icon className={styles.panelExpandoIcon} icon="chevron-down" />
+          <Heading level={5}>{title}</Heading>
+        </button>
+      )}
+      {isExpanded && (
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply firebase-vscode/webviews/components/ui/PanelSection.tsx:32]: I don't think title needs to be part of the condition. It's expanded by default and if it doesn't have a title it doesn't have a title bar with a button that would let anyone ever collapse it.

File: firebase-vscode/src/logger-wrapper.ts (hunk 36 above)
Comment (hsubox76) [reply firebase-vscode/src/logger-wrapper.ts:4]: Tried using LeveledLogMessage but for some reason Typescript gets confused and keeps thinking the return type needs to be the winston Logger and not the Logger interface in src/logger.ts. Went with (...args) => void.

File: firebase-vscode/src/stubs/inquirer-stub.js (hunk 37 above)
Comment (hsubox76) [reply firebase-vscode/src/stubs/inquirer-stub.js:16]: Did it in a somewhat roundabout way since the stubs are written in commonjs and can't import from ts files - could rewrite this in the future so the stubs are written in ts and compiled.

File: firebase-vscode/src/workflow.ts (hunk 39 above)
Comment (hsubox76) [reply firebase-vscode/src/workflow.ts:251]: Separated out the 2 largest ones.

File: firebase-vscode/webviews/components/ui/PanelSection.tsx (hunk 41 above)
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase-vscode/webviews/components/ui/PanelSection.tsx:32]: Without the extra logic, say if the title is initially non-empty and a user collapses a panel. Later if the panel title is changed to empty, the user will be left in a state where they can no longer see anything at all (since there's no title nor expand button).
